### ISSUE TITLE
NextPageLink property null although additional items exist

### ISSUE THREAD

**Issue by ya-alexander** [Azure/azure-libraries-for-net#1056]
Hi,

I am trying to get all groups that exist in an AAD using the following code: 

`var groups = await this._azure.AccessManagement.ActiveDirectoryGroups.ListAsync();`

This call always returns exactly 100 items, although > 200 groups exist in AAD.  So I tried to query the rest of groups using the following code. 

```
var nextPage = await groups.GetNextPageAsync();
do
{
     allGroups.AddRange(nextPage);
} while (nextPage.Any());
```

Unfortunately, the NextPageLink property of `groups` is always null. However, if I trace the call of `this._azure.AccessManagement.ActiveDirectoryGroups.ListAsync()` in Fiddler, the `odata.nextLink` seems to be set correctly:

```
{
    "odata.metadata": "https://graph.windows.net/tenant-guid/$metadata#directoryObjects",
    "odata.nextLink": "directoryObjects/$/Microsoft.DirectoryServices.Group?$skiptoken=X'445370740200010000002A47726F75705F36366364373630312D386537302D343032652D613738612D3933633132393836386538302A47726F75705F36366364373630312D386537302D343032652D613738612D3933633132393836386538300000000000000000000000'",
    "value": [...]
}
```

I also tried to fetch additional groups in Fiddler using the `skiptoken` in a hand-crafted request and everything works as expected in Fiddler.

However, I cannot get to work pagination in my .NET code. I would expect that the `NextPageLink` property is != null if there are more than 100 groups available in AAD. 

Environment:
 - OS: Win10
 - IDE : VS 2019
 - Version of the Library: 1.33.0, but also tried various older versions 

Am I missing something, or could it be that pagination is not working correctly? 

Thanks in advance,
Alex

- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by allenjzhang** [comment]
Looking at the fiddler trace, the nextLink is NOT a valid full path URL. That may be the reason mgmt SDK is unable to retrieve the next page of data.

**Comment by ya-alexander** [comment]
@allenjzhang Thanks for pointing that out. Is there a way I can influence the nextLink to be a full path URL? Am I using the library wrongly, or is this an issue of the underlying service?

**Comment by allenjzhang** [comment]
Underlying service sets the return value. Mm, actually looked at the odata spec, the nextLink may be relative path to the current request replacing the last segment. Some services choose to return the absolute path but some may return relative path.

To help our investigation:
1. Can you confirm you are using .NET fluent library and version?
2. The [ListAsync() call](https://github.com/Azure/azure-libraries-for-net/blob/0ea10a867c2e759bb05a929726cd7b18c20b0a6b/src/ResourceManagement/Graph.RBAC/ActiveDirectoryGroupsImpl.cs) first param is `bool loadAllPages` and is default to `true`. So if you call without setting it to `false`, we may collect everything for you. To confirm this, can you set it to `false` AND try to get a large list to see if paging is now in effect?

Thanks!

**Comment by ya-alexander** [comment]
1. I use the .NET fluent library (Microsoft.Azure.Management.Fluent) version 1.33.0
2. I just tried to call `ListAsync()`, `ListAsync(true)` and `ListAsync(false)`. All three requests resulted in a response containing only 100 items and 

`odata.nextLink=directoryObjects/$/Microsoft.DirectoryServices.Group?$skiptoken=X'445370740200010000002A47726F75705F36366364373630312D386537302D343032652D613738612D3933633132393836386538302A47726F75705F36366364373630312D386537302D343032652D613738612D3933633132393836386538300000000000000000000000'`

I would expect that `ListAsync()` and `ListAsync(true)` return **all** groups of the AAD and the response not containing a `nextLink` (perhaps I am mistaken here). On the other hand, `ListAsync(false)` should return just a page of groups and the `NextPageLink` property should not be null.

Hope that helps.

**Comment by weidongxu-microsoft** [comment]
@ya-alexander

The cause is that the response put the *nextLink* in `odata.nextLink` while .NET SDK is always expecting `nextLink` (which is a deeper issue).

You can try this code as workaround (I am using `ActiveDirectoryUsers`, replace it with `ActiveDirectoryGroups`)
```
List<Microsoft.Azure.Management.Graph.RBAC.Fluent.Models.UserInner> users = new List<Microsoft.Azure.Management.Graph.RBAC.Fluent.Models.UserInner>();
var firstPageResult = azure.AccessManagement.ActiveDirectoryUsers.Inner.ListWithHttpMessagesAsync().Result;
var response = firstPageResult.Response.Content.ReadAsStringAsync().Result;
users.AddRange(firstPageResult.Body);
var deserialized = Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, Object>>(response);
Object nextLink;
bool hasNext = deserialized.TryGetValue("odata.nextLink", out nextLink);
while (hasNext)
{
    var nextPageResult = azure.AccessManagement.ActiveDirectoryUsers.Inner.ListNextWithHttpMessagesAsync(nextLink.ToString()).Result;
    users.AddRange(nextPageResult.Body);
    response = nextPageResult.Response.Content.ReadAsStringAsync().Result;
    deserialized = Newtonsoft.Json.JsonConvert.DeserializeObject<Dictionary<string, Object>>(response);
    hasNext = deserialized.TryGetValue("odata.nextLink", out nextLink);
}
```

De-serialized result is in `firstPageResult.Body`, `nextPageResult.Body`.

**Comment by ya-alexander** [comment]
Thanks! I will give it a try and report back.

**Comment by ya-alexander** [comment]
Just tried it out. Works! The only drawback is that it returns `GroupInner` objects instead of `IActiveDirectoryGroup` objects. So I could not get the members of the according group using `ListMembersAsync()`. Do you have any advice how to achieve that with your sample code?

Thanks!

**Comment by weidongxu-microsoft** [comment]
You can use e.g. (`group` being the `GroupInner` object, `ADGroupInner` class on me?)
```
azure.AccessManagement.ActiveDirectoryGroups.Inner.GetGroupMembersAsync(group.ObjectId).Result;
```

Note this one again is a paged list, and nextLink is also on `odata.nextLink`.

**Comment by weidongxu-microsoft** [comment]
Trying a fix for next version 1.34

**Comment by ya-alexander** [comment]
Thanks! For me the issue is resolved. Looking forward to 1.34.
